File: leetcode/Algorithms/101.SymmetricTree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

public class Solution {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) return true;

        Stack<TreeNode> stack = new Stack<>();
        stack.push(root.left);
        stack.push(root.right);
        while (!stack.isEmpty()) {
            TreeNode a = stack.pop(), b = stack.pop();
            if (a == null && b == null) continue;
            if (a == null || b == null || a.val != b.val) return false;
            stack.push(a.left);
            stack.push(b.right);
            stack.push(b.left);
            stack.push(a.right);
        }

        return true;
    }
}

File: leetcode/Algorithms/101.SymmetricTree/Solution.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

bool isMirror(struct TreeNode* left, struct TreeNode* right) {
    if (left == NULL || right == NULL) return left == right;
    return left->val == right->val && isMirror(left->left, right->right) && isMirror(right->left, left->right);
}

bool isSymmetric(struct TreeNode* root) {
    return root == NULL || isMirror(root->left, root->right);
}
